stock_available_to_promise_release/models/stock_move.py

```diff
@@ -616,27 +700,74 @@ def unrelease(self, safe_unrelease=False):
         moves_to_unrelease.write({"need_release": True})
         impacted_picking_ids = set()
 
+        qty_to_return_per_move = defaultdict(float)
         for move in moves_to_unrelease:
+            rounding = move.product_id.uom_id.rounding
+            # When a move is returned, it is going straight to WH/Stock,
+            # skipping all intermediate zones (pick/pack).
+            # That is why we need to keep track of qty returned along the way.
+            # We do not want to return the same goods at each step.
+            # At a given step (pick/pack/ship), qty to return is
+            # move.product_uom_qty - cancelled_qty_at_step - already returned qties
+            qty_to_unrelease = move.product_qty
+            qty_returned_for_move = 0
             iterator = move._get_chained_moves_iterator("move_orig_ids")
-            moves_to_cancel = self.env["stock.move"]
+            moves_to_cancel_for_move = self.env["stock.move"]
             # backup procure_method as when you don't propagate cancel, the
             # destination move is forced to make_to_stock
             procure_method = move.procure_method
             next(iterator)  # skip the current move
             for origin_moves in iterator:
-                origin_moves = origin_moves.filtered(
+                todo_origin_moves = origin_moves.filtered(
                     lambda m: m.state not in ("done", "cancel")
                 )
-                if origin_moves:
-                    origin_moves = move._split_origins(origin_moves)
-                    impacted_picking_ids.update(origin_moves.mapped("picking_id").ids)
+                qty_cancelled_at_step = 0
+                if todo_origin_moves:
+                    qty_to_split = qty_to_unrelease - qty_returned_for_move
+                    moves_to_cancel = move._split_origins(
+                        todo_origin_moves, qty=qty_to_split
+                    )
```

Comment:
Could it happen that qty_to_split is zero? If so the qty_to_split would be set to the move's product_qty in method _split_origins. I think that would not be correct or?
```suggestion
                    qty_to_split = qty_to_unrelease - qty_returned_for_move
                    if not qty_to_split:
                        continue
                    moves_to_cancel = move._split_origins(
                        todo_origin_moves, qty=qty_to_split
                    )
```

Reply:
i would also rename `qty_to_split` into `qty_to_cancel` for clarity and do that check as first step in the loop. If nothing left to cancel, you can even `break`

```
            for origin_moves in iterator:
                qty_to_cancel = qty_to_unrelease - qty_returned_for_move
                if float_compare(qty_to_cancel, 0, precision_rounding=move.product_uom.rounding) <= 0:
                    break
```